lib/ansible/plugins/cliconf/frr.py

```diff
@@ -0,0 +1,150 @@
+#
+# (c) 2018 Red Hat Inc.
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+---
+author: Ansible Networking Team
+cliconf: frr
+short_description: Use frr cliconf to run command on Free Range Routing platform
+description:
+  - This frr plugin provides low level abstraction api's for
+    sending and receiving CLI commands from devices running frr.
+version_added: "2.8"
+"""
+
+import re
+import json
+
+from ansible.module_utils._text import to_text
+from ansible.module_utils.common._collections_compat import Mapping
+from ansible.module_utils.network.common.utils import to_list
+from ansible.plugins.cliconf import CliconfBase, enable_mode
+
+
+class Cliconf(CliconfBase):
+
+    def get_supported_protocols(self):
+        supported_protocols = {}
+        protocols = ['bgp', 'isis', 'ospf', 'ldp', 'ospf6', 'pim', 'rip', 'ripm', 'zebra']
+        daemons = self.get('show daemons')
+        data = to_text(daemons, errors='surrogate_or_strict').strip()
+
+        for item in protocols:
+            supported_protocols[item] = True if item in data else False
+        return supported_protocols
+
+    def get_device_info(self):
+        device_info = {}
+
+        device_info['network_os'] = 'frr'
+        reply = self.get('show version')
+        data = to_text(reply, errors='surrogate_or_strict').strip()
+
+        match = re.search(r'FRRouting (\S+) \((\S+)\)', data)
+        if match:
+            device_info['network_os_version'] = match.group(1)
+            if match.group(2):
+                device_info['network_os_hostname'] = match.group(2)
+
+        return device_info
+
+    def get_option_values(self):
+        return {
+            'format': ['text'],
+            'output': ['text']
+        }
+
+    def get_device_operations(self):
+        return {
+            'supports_diff_replace': False,
+            'supports_commit': False,
+            'supports_rollback': False,
+            'supports_defaults': False,
+            'supports_onbox_diff': False,
```

Comment:
@NilashishC We can have `get_diff` method for frr which generates diff between running config and candidate config?

Reply:
Added get_diff @trishnaguha